### PR TITLE
[SPARK-39750][SQL] Enable `spark.sql.cbo.enabled` by default

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR enable `spark.sql.cbo.enabled` by default.

### Why are the changes needed?

1. Enable CBO to get better performance, we've enabled it over 3 years.
2. Benchmark related tests also enabled it:https://github.com/apache/spark/blob/e83f8a872a16d4f049cefb1fc445f91cf84443ad/sql/core/src/test/scala/org/apache/spark/sql/TPCBase.scala#L32

### Does this PR introduce _any_ user-facing change?

I don't think this PR could introduce any user-facing changes, we've fixed related bugs before, for example: https://github.com/apache/spark/pull/29894.

### How was this patch tested?

Fix existing tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 1]
```diff
@@ -353,48 +353,65 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
...
+                   |CREATE TABLE $table (col1 int, col2 int)
+                   |USING PARQUET
+                   |PARTITIONED BY (col2)
+                   |LOCATION '${dir.toURI}'""".stripMargin)
+              sql(s"ANALYZE TABLE $table COMPUTE STATISTICS")
+              spark.table(table)
+              if (!cboEnabled) {
+                assert(getTableFromCatalogCache(table).stats.sizeInBytes === 0)
+              } else {
+                assert(getTableFromCatalogCache(table).stats.sizeInBytes === 1)
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala:400]: The size is 1, not 0:
https://github.com/apache/spark/blob/ddc61e62b9af5deff1b93e22f466f2a13f281155/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala#L116-L118

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala [hunk 2]
```diff
@@ -58,6 +58,7 @@ case class FilterEstimation(plan: Filter) extends Logging {
         rowsAfterFilter = filteredRowCount)
     }
     val filteredSizeInBytes: BigInt = getOutputSize(plan.output, filteredRowCount, newColStats)
+      .min(plan.child.stats.sizeInBytes)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala:61]: To fix the test error:
image

File: sql/core/src/test/scala/org/apache/spark/sql/InjectRuntimeFilterSuite.scala [hunk 3]
```diff
@@ -209,11 +209,13 @@ class InjectRuntimeFilterSuite extends QueryTest with SQLTestUtils with SharedSp
     // `MergeScalarSubqueries` can duplicate subqueries in the optimized plan and would make testing
     // complicated.
     conf.setConfString(SQLConf.OPTIMIZER_EXCLUDED_RULES.key, MergeScalarSubqueries.ruleName)
+    conf.setConf(SQLConf.CBO_ENABLED, false)
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/InjectRuntimeFilterSuite.scala:212]: To fix the test error:
image

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/DynamicPartitionPruningHiveScanSuite.scala [hunk 4]
```diff
@@ -41,6 +42,16 @@ abstract class DynamicPartitionPruningHiveScanSuiteBase
       case _ => Nil
     }
   }
+
+  override protected def beforeAll(): Unit = {
+    super.beforeAll()
+    conf.setConf(SQLConf.CBO_ENABLED, false)
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/DynamicPartitionPruningHiveScanSuite.scala:48]: To fix the test error:

**Comment by wangyum** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
cc @sigmod @maryannxue @viirya

**Review by dongjoon-hyun — COMMENTED** [review]
Although I support your PR, are you sure about the following, @wangyum ?

> I don't think this PR could introduce any user-facing changes,

Please update SQL migration guide, please.

**Comment by dongjoon-hyun** [comment]
cc @sunchao , @viirya , @huaxingao

**Comment by sunchao** [comment]
@wangyum do you have any numbers on the performance gain from this?

**Review by viirya — COMMENTED** [review]
Adaptive execution is now enabled by default. What the interaction could be if we enable CBO by default too? Do you have some idea? @wangyum

**Comment by wangyum** [comment]
> @wangyum do you have any numbers on the performance gain from this?

Will update the benchmark results later.

**Comment by wangyum** [comment]
@viirya Here is an example. AQE can be converted to broadcast join, but the performance is worse than directly planned to broadcast join.

```scala
import org.apache.spark.benchmark.Benchmark
val numRows = 1024 * 1024 * 25
spark.sql(s"CREATE TABLE t1 using parquet AS SELECT id AS a, id AS b, id AS c, id AS d, id AS e FROM range(${numRows}L)")
spark.sql(s"CREATE TABLE t2 using parquet AS SELECT id AS a, id AS b, id AS c, id AS d, id AS e FROM range(${numRows}L)")

spark.sql("ANALYZE TABLE t1 COMPUTE STATISTICS FOR ALL COLUMNS")
spark.sql("ANALYZE TABLE t2 COMPUTE STATISTICS FOR ALL COLUMNS")

val benchmark = new Benchmark(s"Benchmark SPARK-39750", numRows, minNumIters = 5)
Seq(false, true).foreach { enabled =>
  val name = s"CBO is ${if (enabled) "Enabled" else "Disabled"}"
  benchmark.addCase(name) { _ =>
    withSQLConf("spark.sql.cbo.enabled" -> s"$enabled") {
      spark.sql(s"SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND t2.b < 10000").write.format("noop").mode("Overwrite").save()
    }
  }
}
benchmark.run()
```

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_281-b09 on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Benchmark SPARK-39750:                    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
CBO is Disabled                                    5277           5570         195          5.0         201.3       1.0X
CBO is Enabled                                     1378           1582         166         19.0          52.6       3.8X
```

Default | Enable CBO
-- | --
image | image

File: sql/core/src/test/scala/org/apache/spark/sql/InjectRuntimeFilterSuite.scala (hunk 3 above)
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/InjectRuntimeFilterSuite.scala:212]: Seems this line is to keep consistent with the previous behavior, but should we add some TODO for this change? For example, spilt this Suites into `withCBOEnabled` and `withCBODisabled`?

I found many suites in this class will fail when `CBOEnabled`.

**Comment by wangyum** [comment]
TPC-DS q47:
Default | Enable CBO
-- | --
image | image

**Comment by dongjoon-hyun** [comment]
Could you add some description about some requirements or assumptions. For example, what happens when the statistics are incorrect?

**Comment by c21** [comment]
This is a great move, thanks @wangyum for starting this! I think CBO and AQE could work well together, with CBO handles query without shuffle in a better way, and AQE handles query with shuffle in a better way. With more accurate AQE statistics, CBO optimizer could also do a better job for query planning.

But I shared similar question with @dongjoon-hyun, when the statistics is out of date to become inaccurate, current CBO has no way to check it AFAIK. It would be great if we can add more check / work around for it in 3.4 release to either (1).make statistics up-to-date, or (2).automatically disable CBO if statistics is stale.

**Comment by mridulm** [comment]
+CC @zoelin7 FYI

**Pull request closed without merging**